### ISSUE TITLE
int and char do not compare correctly?

### ISSUE THREAD

**Issue by oznetmaster** [nunit/nunit#524]
char c = '\u0000';
Assert.AreEqual (0, c) fails.

Yet, if you cast c to an int, it has the value of 0.

Shouldn't 0 and '\u0000' be equal?

**Comment by oznetmaster** [comment]
I have been told that this does work correctly under NUnit v2??

**Comment by CharliePoole** [comment]
We should check under NUnit 2.6.4. If it works there, it should be made to work in 3.0 as well.

**Comment by oznetmaster** [comment]
Even if it does not work there, we need to consider if it should work.  It seems that char perhaps should be a valid Numeric, since it can be used in any numeric expression.

**Comment by CharliePoole** [comment]
What happens with these Assert.That(c, Is.EqualTo(0)) ?

This change won't fix it because it only deals with the classic assert form rather than the underlying definition of equality in NUnit. One solution would be for us to treat char as a numeric in Numerics.cs as we do with Byte and SByte. But chars can have various representations (single-byte, multi-byte, unicode, etc.) so we didn't. What would be the implications if char were numeric?

Honestly, I'm not sure what is the right thing to do here.

**Comment by constructor-igor** [comment]
The 2 tests fail (NUnit v3)
```c#
    [TestFixture]
    public class NUnitIssue524
    {
        [Test]
        public void AssertAreEqual()
        {
            char c = '\u0000';
            Assert.AreEqual(0, c);
        }
        [Test]
        public void AssertThat()
        {
            char c = '\u0000';
            Assert.That(c, Is.EqualTo(0));
        }
    }
```
![image](https://cloud.githubusercontent.com/assets/1849690/7227339/8edae27e-e758-11e4-8f7f-e47e9e9cff15.png)

![image](https://cloud.githubusercontent.com/assets/1849690/7227348/99e351d8-e758-11e4-9e41-89a0a4f0d998.png)

**Comment by constructor-igor** [comment]
NUnit v2 report:

Assert.AreEqual passes
```c#
        [Test]
        public void AssertAreEqual()
        {
            char c = '\u0000';
            Assert.AreEqual(0, c);
        }
```

Assert.That fails
```c#
        [Test]
        public void AssertThat()
        {
            char c = '\u0000';
            Assert.That(c, Is.EqualTo(0));
        }
```
![image](https://cloud.githubusercontent.com/assets/1849690/7227421/1645ea7e-e759-11e4-8098-ac2d2c1452b2.png)

**Comment by CharliePoole** [comment]
It's understandable that one passes and the other fails in V2. That's because there's a cast to int involved in Assert.AreEqual, so that two ints actually get compared. Ironically, it means that this comparison was permitted before NUnit 2.4 and has been partially 'broken' ever since Assert.That was introduced.

We could fix it in NUnitEqualityComparer if we decided that passing the comparison is desired. I'm not convinced and in any case I think we need to decide on the  scope of what we want to "work" here. For example, should two characters from different character sets be equal if they have the same numeric value? I seriously doubt it. A char compared to an int is a different matter though, since it implies the user expected a conversion.

I'm changing this to status:review until we decide. Possible outcomes seem to be:

1. Leave it as is. User must do an explicit cast to int. Seems reasonable to me if documented.

2. Make char to int and int to char comparison a special case, which implies a cast. Also seems reasonable.

3. Always treat char as a numeric. This would make the numeric value of the char be used for all equality comparisons. The byte representation and character set would not matter. I don't like this myself.

**Comment by rprouse** [comment]
Since there is an implicit cast between char and numeric types in C#, I would support a variation on 3.

For example, I can do the following without casting,

```C#
char c = '\u0000';
int i = 0;
if (c == i) Console.WriteLine("They are equal");
```

I think that anything that evaluates to true in an `if ( x == y )` should also evaluate to true with both forms of NUnit asserts.

I don't however think we should do it with overloads. I think we should modify `Numerics.IsFixedPointNumeric` to add `char` and then modify `NUnitComparer` and `NUnitEqualityComparer` to do a specific character comparison if both arguments are char before it falls through to the numeric comparisons. I think this will address @CharliePoole's concerns with number 3?

Can anyone see any holes or misses in this idea?

**Comment by CharliePoole** [comment]
I like the solution as well as the general principle that we should allow equality where the operator allows it. Yes, adding a special case for char-char comparisons would resolve the issues I raised.

**Comment by oznetmaster** [comment]
It is what I suggested as well.

---Sent from my iPad

On April 20, 2015 at 11:53:43 PM GMT+8, Rob Prouse <notifications@github.com> wrote:Since there is an implicit cast between char and numeric types in C#, I would support a variation on 3.  For example, I can do the following without casting,  char c = '\u0000'; int i = 0; if (c == i) Console.WriteLine("They are equal");   I think that anything that evaluates to true in an if ( x == y )  should also evaluate to true with both forms of NUnit asserts.  I don't however think we should do it with overloads. I think we should modify Numerics.IsFixedPointNumeric  to add char  and then modify NUnitComparer  and NUnitEqualityComparer  to do a specific character comparison if both arguments are char before it falls through to the numeric comparisons. I think this will address @CharliePoole's concerns with number 3?  Can anyone see any holes or misses in this idea?  —Reply to this email directly or view it on GitHub.

**Comment by oznetmaster** [comment]
I agree

**Comment by CharliePoole** [comment]
Should have been closed by PR #657
